### PR TITLE
Update types to use generics & add a TS test for react-redux

### PR DESCRIPTION
**Pull request by kale-stew**
## Description

Rel: #69

This PR introduces a test for react-redux's `useSelector` hook to access state. Ideally, we should also confirm the # of overlapping values to make sure we're using the hook as expected, but I didn't want to get too into the weeds of wrapping another test around this one.

Would appreciate feedback / input on how I can enhance this test.

## Checklist:

- [x] All tests are passing
- [x] Type definitions, if updated, pass both `test-ts-defs` and `test-ts-usage`
- [x] Benchmark performance has not significantly decreased
- [x] Bundle size has not been significantly impacted
- [x] ~The bundle size badge has been updated to reflect the new size~ **N/A**, still at 369

### CODE REVIEW

File: test/typescript/sample-usage.tsx [hunk 1]
```diff
@@ -14,26 +15,44 @@ const testArr: ITodo[] = [
...
 }
 
-class TestContainer extends React.Component {
+type IContainerState = { completed: ITodo[] };
+
+class TestContainer extends React.Component<{}, IContainerState> {
   render() {
-    return testArr.map((item) => <TestChild key={item.id} todo={item} />);
+    const numberOfOverlappingItems: number = useSelector(
+      (state) => this.state.completed.length
```
Comment (chrisbolin) [inline test/typescript/sample-usage.tsx:43]: we need to call `useSelector` as the issue reporter does here: https://github.com/FormidableLabs/react-fast-compare/issues/61

The key is the second arg, the `equalityFn`.

Hopefully the test will then _fail_. It should fail in the same way #61 does. Once we've proved that failure, we can modify the types and all the tests should pass.

Comment (chrisbolin) [reply test/typescript/sample-usage.tsx:43]: I believe we will need a redux container, as the docs show here: https://react-redux.js.org/api/hooks. The Provider sets the _type_ of the state, which is very important to us :)

File: test/typescript/sample-usage.tsx [hunk 2]
```diff
@@ -14,26 +15,44 @@ const testArr: ITodo[] = [
...
     const todo = this.props.todo;
+
     return <div>{todo.text}</div>;
   }
 }
 
-class TestContainer extends React.Component {
+type IContainerState = { completed: ITodo[] };
+
+class TestContainer extends React.Component<{}, IContainerState> {
```
Comment (chrisbolin) [inline test/typescript/sample-usage.tsx:40]: Let's have a completely new component (or set of components, if necessary) for this test. I don't want our components to do "double duty" by trying to test multiple things. These components sure be single-minded, focused on testing our types against react-redux

File: package.json [hunk 3]
```diff
@@ -8,12 +8,12 @@
...
     "benchmark": "node benchmark",
     "eslint": "eslint \"*.js\" benchmark test",
-    "tslint": "eslint test/typescript/sample-usage.tsx",
+    "tslint": "eslint test/typescript/*.tsx",
     "test-browser": "karma start test/browser/karma.conf.js",
     "test-browser-ie": "karma start test/browser/karma.conf.ie.js",
     "test-node": "mocha \"test/node/*.spec.js\"",
     "test-node-cov": "nyc mocha \"test/node/*.spec.js\"",
-    "test-ts-usage": "tsc --esModuleInterop --jsx react --noEmit test/typescript/sample-usage.tsx",
+    "test-ts-usage": "tsc --esModuleInterop --jsx react --noEmit test/typescript/*.tsx",
```
Comment (chrisbolin) [inline package.json:16]: i like this approach!

File: test/typescript/sample-usage.tsx [hunk 4]
```diff
@@ -14,14 +14,16 @@ const testArr: ITodo[] = [
   { text: 'mountain', id: '11' },
   { text: 'air', id: '8' },
   { text: 'plants', id: '9' },
+  { text: 'air', id: '8' },
```
Comment (chrisbolin) [inline test/typescript/sample-usage.tsx:17]: might as well just remove any changes to this file so we can keep the review smaller

File: test/typescript/sample-react-redux-usage.tsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      return state;
+  }
+};
+
+const store = createStore(overlap, ['mountain']);
+
+class TestContainer extends React.Component<{}, IContainerState> {
+  render() {
+    // to interact with the store
+    // store.dispatch({ type: 'TYPE', text: 'some-text' })
```
Comment (chrisbolin) [inline test/typescript/sample-react-redux-usage.tsx:57]: we shouldn't need to dispatch or anything like that. just use the selector hook in the render

File: test/typescript/sample-react-redux-usage.tsx [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  { text: 'plants', id: '9' },
+  { text: 'air', id: '8' },
+  { text: 'air', id: '8' },
+];
+
+type IChildProps = {
+  item: IItem;
+};
+
+class TestChild extends React.Component<IChildProps> {
```
Comment (chrisbolin) [inline test/typescript/sample-react-redux-usage.tsx:30]: we can totally remove the child from this test. _all_ we need is really state, provider, and a component that inherits from the provide and uses `useSelector` _with react-fast-compare_.

File: package.json (hunk 3 above)
Comment (kale-stew) [reply package.json:16]: AppVeyor doesn't seem to 😕 

```sh
$ builder concurrent --buffer eslint tslint test-ts-usage test-ts-defs test-node-cov test-browser-ie
error TS6053: File 'test/typescript/*.tsx' not found.
```

**Pull request merged**